### PR TITLE
[ML] Assorted tidy ups

### PR DESCRIPTION
**Pull request by tveasey**
This addresses a variety of sensible clang-tidy suggestions.

Principally:
1. Migrating from `std::bind` to lambdas,
2. Addressing implicit conversions of ints and pointers to bools,
3. Matching function declaration and definition parameter names,
4. Not using lower case explicit integer type qualifiers,
5. Using `empty()` rather than 'size() == 0`,
6. Using explicit constructors,
7. Preferring range based for loops,
8. Preferring auto when initialising from a cast.

These are not fully addressed in the code base. I focused mainly on the `maths` library. We can gradually address more cases in other PR. Whilst making these changes I noticed in various places we were also silently swallowing errors restoring types in constructors. In these, cases we now set the traverser state as bad.

Finally, we can now remove pre 6.3 state restoration from main. Code and associated tests for these upgrade paths have been removed.

### CODE REVIEW

File: lib/maths/time_series/CTimeSeriesDecomposition.cc [hunk 1]
```diff
@@ -127,59 +121,38 @@ bool CTimeSeriesDecomposition::acceptRestoreTraverser(const common::SDistributio
...
-                        &CComponents::acceptRestoreTraverser, &m_Components,
-                        std::cref(params), m_LastValueTime, std::placeholders::_1)))
+                    traverser.traverseSubLevel([this](auto& traverser_) {
+                        return m_CalendarCyclicTest.acceptRestoreTraverser(traverser_);
+                    }))
+            RESTORE(COMPONENTS_6_3_TAG, traverser.traverseSubLevel([&](auto& traverser_) {
+                return m_Components.acceptRestoreTraverser(params, traverser_);
+            }))
         }
-    } else {
```
Comment (droberts195) [inline lib/maths/time_series/CTimeSeriesDecomposition.cc:165]: Maybe return false if neither supported version tag is found?

File: lib/maths/time_series/CTimeSeriesModel.cc [hunk 2]
```diff
@@ -1268,59 +1261,35 @@ bool CUnivariateTimeSeriesModel::acceptRestoreTraverser(const common::SModelRest
...
                 m_AnomalyModel = std::make_unique<CTimeSeriesAnomalyModel>(),
-                traverser.traverseSubLevel(std::bind(
-                    &CTimeSeriesAnomalyModel::acceptRestoreTraverser,
-                    m_AnomalyModel.get(), std::cref(params), std::placeholders::_1)),
+                traverser.traverseSubLevel([&](auto& traverser_) {
+                    return m_AnomalyModel->acceptRestoreTraverser(params, traverser_);
+                }),
                 /**/)
         }
-    } else {
```
Comment (droberts195) [inline lib/maths/time_series/CTimeSeriesModel.cc:1295]: Else return false?

File: lib/maths/time_series/CTimeSeriesModel.cc [hunk 3]
```diff
@@ -2696,59 +2662,38 @@ bool CMultivariateTimeSeriesModel::acceptRestoreTraverser(const common::SModelRe
...
```
Comment (droberts195) [inline lib/maths/time_series/CTimeSeriesModel.cc:2724]: Else return false?

**Comment by droberts195** [comment]
Also, in case you didn't see it:

> A format error has been detected within the following files:
> `lib/api/unittest/CRestorePreviousStateTest.cc`

File: lib/maths/time_series/CTimeSeriesDecomposition.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/time_series/CTimeSeriesDecomposition.cc:165]: Done in [ca10699](https://github.com/elastic/ml-cpp/pull/2221/commits/ca106994958c0af9ea88d7b2bb9d008293d1281c).

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**